sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,301 @@
+#include <sycl/exception.hpp>
+#include <sycl/ext/oneapi/annotated_arg/properties.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_HOST_NOT_SUPPORTED(Op)
+#else
+#define __SYCL_HOST_NOT_SUPPORTED(Op)                                          \
+  throw sycl::exception(                                                       \
+      sycl::make_error_code(sycl::errc::feature_not_supported),                \
+      Op " is not supported on host device.");
+#endif
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __OPENCL_GLOBAL_AS__ __attribute__((opencl_global))
+#ifdef __ENABLE_USM_ADDR_SPACE__
+#define __OPENCL_GLOBAL_DEVICE_AS__ __attribute__((opencl_global_device))
+#define __OPENCL_GLOBAL_HOST_AS__ __attribute__((opencl_global_host))
+#else
+#define __OPENCL_GLOBAL_DEVICE_AS__ __attribute__((opencl_global))
+#define __OPENCL_GLOBAL_HOST_AS__ __attribute__((opencl_global))
+#endif // __ENABLE_USM_ADDR_SPACE__
+#define __OPENCL_LOCAL_AS__ __attribute__((opencl_local))
+#define __OPENCL_CONSTANT_AS__ __attribute__((opencl_constant))
+#define __OPENCL_PRIVATE_AS__ __attribute__((opencl_private))
+#else
+#define __OPENCL_GLOBAL_AS__
+#define __OPENCL_GLOBAL_DEVICE_AS__
+#define __OPENCL_GLOBAL_HOST_AS__
+#define __OPENCL_LOCAL_AS__
+#define __OPENCL_CONSTANT_AS__
+#define __OPENCL_PRIVATE_AS__
+#endif
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+namespace detail {
+
+// Type-trait for checking if a type defines `operator[]`.
+template <typename T, typename = void>
+struct HasSubscriptOperator : std::false_type {};
+
+template <typename T>
+struct HasSubscriptOperator<
+    T, typename std::enable_if_t<
+           !std::is_void<decltype(std::declval<T>().operator[](0))>::value>>
+    : std::true_type{};
+
+} // namespace detail
+
+// Deduction guide
+template <typename T, typename... Args>
+annotated_arg(T, Args... args)
+    -> annotated_arg<T, detail::properties_t<Args...>>;
+
+template <typename T, typename... Args>
+annotated_arg(T, properties<std::tuple<Args...>>)
+    -> annotated_arg<T, detail::properties_t<Args...>>;
+
+template <typename T, typename old, typename... ArgT>
+annotated_arg(annotated_arg<T, old>, properties<std::tuple<ArgT...>>)
+    -> annotated_arg<
+        T, detail::merged_properties_t<old, detail::properties_t<ArgT...>>>;
+
+template <typename T, typename PropertyListT = detail::empty_properties_t>
+class annotated_arg {
+  // This should always fail when instantiating the unspecialized version.
+  static_assert(is_property_list<PropertyListT>::value,
+                "Property list is invalid.");
+};
+
+// Partial specialization for pointer type
+template <typename T, typename... Props>
+class __SYCL_SPECIAL_CLASS
+__SYCL_TYPE(annotated_arg) annotated_arg<T *, detail::properties_t<Props...>> {
+  using property_list_t = detail::properties_t<Props...>;
+  __OPENCL_GLOBAL_AS__ T *obj;
```

Comment:
Not a requirement, but you could add a `using decorated_pointer_t = __OPENCL_GLOBAL_AS__ T *` here and use that the places where it's relevant. Slightly shorter and feel free to use another name too if you like.